numpy/core/arrayprint.py

```diff
@@ -1209,6 +1211,31 @@ def array_repr(arr, max_line_width=None, precision=None, suppress_small=None):
             lf = '\n'+' '*len(class_name + "(")
         return "%s(%s,%sdtype=%s)" % (class_name, lst, lf, typename)
 
+
+    if issubclass(arr.dtype.type, flexible):
+        if arr.dtype.names:
+            typename = "%s" % str(arr.dtype)
+        else:
+            typename = "'%s'" % str(arr.dtype)
+    else:
+        typename = arr.dtype.name
+        # quote typenames which can't be represented as python variable names
+        if typename and not (typename[0].isalpha() and typename.isalnum()):
+            typename = "'%s'" % typename
+
+    prefix = "{}({},".format(class_name, lst)
+    newline_ind = prefix.rfind('\n')
+    last_line_len = len(prefix) - (0 if newline_ind == -1 else newline_ind)
+    if max_line_width is None:
+        max_line_width = _format_options['linewidth']
+
+    if last_line_len + len(typename) + len(' dtype=)') > max_line_width:
+        suffix = "\n{}dtype={})".format(' '*len(class_name + "("), typename)
+    else:
+        suffix = " dtype={})".format(typename)
```

Comment:
Too much duplication of `dtype=` here for my liking. If you assembled this in smaller pieces, you wouldn't need to keep building the same piece